app/lib/search/index_simple.dart

```diff
@@ -71,66 +71,59 @@ class SimplePackageIndex implements PackageIndex {
 
   @override
   Future<PackageSearchResult> search(SearchQuery query) async {
-    final Map<String, double> total = <String, double>{};
-    void addAll(Map<String, double> scores, double weight) {
-      scores?.forEach((String url, double score) {
-        if (score != null) {
-          final double prev = total[url] ?? 0.0;
-          total[url] = prev + score * weight;
-        }
-      });
-    }
-
-    addAll(_nameIndex.search(query.text), 0.70);
-    addAll(_descrIndex.search(query.text), 0.10);
-    addAll(_readmeIndex.search(query.text), 0.05);
-
-    if ((query.text == null || query.text.isEmpty) &&
-        query.packagePrefix != null) {
-      addAll(_nameIndex.search(query.packagePrefix), 0.8);
+    // do text matching
+    final Score textScore = _searchText(query.text, query.packagePrefix);
+
+    // The set of urls to filter on.
+    final Set<String> urls =
+        textScore?.getKeys()?.toSet() ?? _documents.keys.toSet();
+
+    // filter on package prefix
+    if (query.packagePrefix != null) {
```

Comment:
For future optimizations: There might be possibilities where one can avoid constructing the big maps in `Score` if we know only a small subset of keys will survive.

i.e. Instead of building up big datastructures and removing from them later, we could try filtering early on and construct smaller datastructures.

Reply:
Good point, will address it in a follow-up.